File: ref/code/L02/U2021jans2.java
package com.deltacoding.usaco.s2020_2021.jan2021.silver;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class U2021jans2 {

	public static void main(String[] args) throws IOException {
		BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(in.readLine());
		int N = Integer.parseInt(st.nextToken());
		int Q = Integer.parseInt(st.nextToken());
		int[] prefix = new int[N + 2];
		int[] suffix = new int[N + 2];
		String s = in.readLine();

        // prefix sum
		for (int c = 'A'; c <= 'Z'; c++) {
			int cnt = 0;
			boolean brushdown = false;
			for (int i = 1; i <= N; i++) {
				if (s.charAt(i-1) == c) {
					if (!brushdown) {
						brushdown = true;
						cnt++;
					}
				}
				if (s.charAt(i-1) < c) {
					brushdown = false;
				}
				prefix[i] += cnt;
			}
		}

		// suffix sum
		for (int c = 'A'; c <= 'Z'; c++) {
			int cnt = 0;
			boolean brushdown = false;
			for (int i = N; i > 0; i--) {
				if (s.charAt(i-1) == c) {
					if (!brushdown) {
						cnt++;
						brushdown = true;
					}
				}
				if (s.charAt(i-1) < c) {
					brushdown = false;
				}
				suffix[i] += cnt;
			}
		}

		for (int i = 1; i <= Q; i++) {
			st = new StringTokenizer(in.readLine());
			int l = Integer.parseInt(st.nextToken());
			int r = Integer.parseInt(st.nextToken());
			System.out.println(prefix[l - 1] + suffix[r + 1]);
		}
	}
}

File: ref/code/L02/U2021jans2.cpp
#include <bits/stdc++.h>
using namespace std;

int N, Q, prefix[100002], suffix[100002];
int main() {
    ios::sync_with_stdio(false);
    cin.tie(0), cout.tie(0);
    cin >> N >> Q;
    string s;
    cin >> s;
    for (int c = 'A'; c <= 'Z'; c++) {
        int cnt = 0;
        bool brushdown = false; // false表示当前刷子提起状态
        for (int i = 1; i <= N; i++) {
            if (s[i-1] == c) {
                if (!brushdown) {
                    cnt++;
                    brushdown = true;
                }
            }
            if (s[i - 1] < c) {
                brushdown = false;
            }
            prefix[i] += cnt; // 26种颜色的前缀和累加
        }

        cnt = 0;
        brushdown = false;
        for (int i = N; i > 0; i--) {
            if (s[i - 1] == c) {
                if (!brushdown) {
                    cnt++;
                    brushdown = true;
                }
            }
            if (s[i - 1] < c) {
                brushdown = false;
            }
            suffix[i] += cnt;
        }
    }
    for (int i = 1; i <= Q; i++) {
        int a, b;
        cin >> a >> b;
        cout << prefix[a - 1] + suffix[b + 1] << endl;
    }
    return 0;
}
